core/src/main/java/org/elasticsearch/search/collapse/CollapseBuilder.java

```diff
@@ -52,12 +60,28 @@
     static {
         PARSER.declareString(CollapseBuilder::setField, FIELD_FIELD);
         PARSER.declareInt(CollapseBuilder::setMaxConcurrentGroupRequests, MAX_CONCURRENT_GROUP_REQUESTS_FIELD);
-        PARSER.declareObject(CollapseBuilder::setInnerHits,
-            (p, c) -> InnerHitBuilder.fromXContent(c), INNER_HITS_FIELD);
+        PARSER.declareField((parser, builder, context) -> {
+            XContentParser.Token currentToken = parser.currentToken();
+            if (currentToken == XContentParser.Token.START_OBJECT) {
+                // backwards compatibility for when only a single inner hit was allowed
```

Comment:
We'll keep this format around not only for backward compatibility, the array form is just another format.